### PR TITLE
Generate a hard reference to "caml_startup" in Asmlink

### PR DESCRIPTION
**Pull request by ghost**
Following a discussion on caml-devel, this PR makes sure that object files generated by `-output-complete-obj` always contain `caml_startup`.

We have tests for `-output-complete-obj` in Dune and they test all the shared/dynamic and byte/native combinations. On FreeBSD, most of them fail with master and they all pass with this PR.

### CODE REVIEW

**Comment by shindere** [comment]
Thanks a lot for this PR!

While going through the patch, I was wondering whether it woulnd't be
better to call force_linking_of_startup conditionnally, e.g. only when
!Clflags.output_complete_object is true. Any opinion?

Also: if you so wish, you could re-enable the output-complete-obj test
in the testsuite and make sure it passes. If you do so, beware that it
will probably not pass on Windows so youmay have to keep it disabled in
this platform as is done for the bytecode test.

If you try to re-enable the test, you may want to use precheck to make
sure where it passes and fails.

File: asmcomp/asmlink.ml [hunk 1]
```diff
@@ -203,6 +203,9 @@ let scan_file obj_name tolink = match read_file obj_name with
 
 (* Second pass: generate the startup file and link it with everything else *)
 
+let force_linking_of_startup ppf =
+  Asmgen.compile_phrase ppf (Cmm.Cdata([Cmm.Cglobal_symbol "caml_startup"]))
+
```
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: Like @shindere I'm tempted to have this (or maybe its two calls) protected by `if !Clflags.output_complete_object then ...`, if only to document-in-code that this forced linking is required only in that case.

**Comment by shindere** [comment]
Xavier Leroy (2018/04/27 06:05 -0700):
> xavierleroy commented on this pull request.
> 
> 
> 
> > @@ -203,6 +203,9 @@ let scan_file obj_name tolink = match read_file obj_name with
>  
>  (* Second pass: generate the startup file and link it with everything else *)
>  
> +let force_linking_of_startup ppf =
> +  Asmgen.compile_phrase ppf (Cmm.Cdata([Cmm.Cglobal_symbol "caml_startup"]))
> +
> 
> Like @shindere I'm tempted to have this (or maybe its two calls)
> protected by `if !Clflags.output_complete_object then ...`, if only to
> document-in-code that this forced linking is required only in that
> case.

It would also avoid introducing a change in the generated code in all
the cases where such a change is actually not required.

**Comment by ghost** [comment]
I added the condition and enabled the `output-complete-obj` test. According to precheck, the test succeeds everywhere (including Windows) except on OSX. The error on OSX is the same with and without this PR:

```
$ gcc -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE -Wl,-no_compact_unwind -I/Users/jdimino/ocaml/byterun -o test.ml_stub.exe test.ml.exe.o -lpthread test.ml_stub.c
Undefined symbols for architecture x86_64:
  "_caml_code_area_end", referenced from:
      _segv_handler in test.ml.exe.o
  "_caml_code_area_start", referenced from:
      _segv_handler in test.ml.exe.o
  "_caml_startup", referenced from:
      _main in test-75e20c.o
     (maybe you meant: _caml_startup__frametable, _caml_startup__data_end , _caml_startup__data_begin , _caml_startup__code_end , _caml_startup__code_begin , _caml_startup_aux )
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by ghost** [comment]
Actually it's just that the `.globl _caml_startup` wasn't enough to generate a reference to `caml_startup`. I changed the code to generate a global symbol named `force_link_of_caml_startup` that references `caml_startup` instead. It works when I test manually on OSX, I'm running precheck again with the new code.

**Comment by ghost** [comment]
precheck is happy. @shindere @xavierleroy This PR is ready for another round of review

File: asmcomp/asmlink.ml [hunk 2]
```diff
@@ -203,6 +203,14 @@ let scan_file obj_name tolink = match read_file obj_name with
...
 (* Second pass: generate the startup file and link it with everything else *)
 
+let force_linking_of_startup ppf =
+  let open Cmm in
+  let name = "force_link_of_caml_startup" in
+  Asmgen.compile_phrase ppf
+    (Cdata ([Cglobal_symbol name;
+             Cdefine_symbol name;
+             Csymbol_address "caml_startup"]))
+
```
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: I suspect `Csymbol_address "caml_startup"` is enough, there is no need to give this piece of data a label `force_link_of_caml_startup`.  But you are correct that a mere declaration `Cglobal_symbol "caml_startup"` is not enough, putting the symbol value into the data section is needed to convince the linker.

**Review by xavierleroy — COMMENTED** [review]
Looks good to me, with a possible simplification (see below).

**Comment by shindere** [comment]
Jérémie Dimino (2018/04/30 09:08 -0700):
> precheck is happy. @shindere @xavierleroy This PR is ready for another
> round of review

Thanks!

I just wanted to make clear here that all the tests are currently
protected by the "libunix" test. This means that the tests will run only
on the platforms where libunix (rather than lubwin32unix) is compiled,
i.e. all the platforms except MinGW* and MSVC*. Precheck does succeed on
these platfoms because the tests are skipped.

I am not saying this PR should necessarily fix this, just wanted to make
sure the current state of affairs is clear to everybody.

As mentionned on caml-devel, for the tests to pass (and the
-output-complete-obj option to actually work on MSVC), the hard-coded
occurrences of -lcamlrunin bytecomp/bytelink.ml should probably be
replaced by something more portable, but it's not obvious how to do that
because Microsoft linker does not have an equivalent to the -lshortcut
stuf (@dra27, @nojb?)).

One possibility I had in mind consists in changing the interface to
call_linker so that this function accepts a list of libraries as
argument. It owuld then be responsible in passing them to the linker in
a way that makes sense on that particular platform.

**Comment by ghost** [comment]
Ah, I see. It wasn't clear to me that this was a guard, I thought it was just a title. It would be good to get this to work on Windows indeed.

Regarding this PR, I did the simplification and re-run precheck, so it's good to merge.

**Comment by shindere** [comment]
Jérémie Dimino (2018/05/01 03:07 -0700):
> Ah, I see. It wasn't clear to me that this was a guard, I thought it
> was just a title.

Well it's both actually :)

> It would be good to get this to work on Windows indeed.

Yes!

> Regarding this PR, I did the simplification and re-run precheck, so
> it's good to merge.

Thanks! IMO feel free to merge whenever you want. I guess it should be
cherry-picked to the 4.07 branch, too.

**Comment by shindere** [comment]
Jérémie Dimino (2018/05/01 03:07 -0700):
> Merged #1740.

Oops, I'm arriving too late. :)

Only the cherry-pick remains, I think.

**Comment by ghost** [comment]
Agreed, @damiendoligez I cherry-picked this in 4.07 given that this is strictly a bug-fix.

**Comment by dra27** [comment]
@shindere - FlexDLL 0.36 includes an heuristic to convert -lfoo automatically for the MSVC ports (see https://github.com/alainfrisch/flexdll/pull/32). Is that relevant to this scenario?

**Comment by shindere** [comment]
David Allsopp (2018/05/01 13:14 +0000):
> @shindere - FlexDLL 0.36 includes an heuristic to convert -lfoo
> automatically for the MSVC ports (see
> https://github.com/alainfrisch/flexdll/pull/32).

Oh that's good to know. Thanks!

> Is that relevant to this scenario?

yes, in that it helps me to understand the problem.

AIUI, the -l options are expected to be processed during the stage where
the object file is built by the -output-complete-obj command-line
option. But since, at the moment, flexlink is not called at this stage,
this does not happen and Microsoft's linker receives these -l options it
does not know what to do with. It still builds the object file, though,
but without any library-related information. When this object file is finally
passed to flexlink to build an actual executable, the library information
is missing and this simply can't work.

I don't know how all this could / should be fixed.

**Pull request merged**